soccer/gameplay/plays/offense/adaptive_formation.py

```diff
@@ -0,0 +1,276 @@
+        self.add_transition(AdaptiveFormation.State.passing,
+                            AdaptiveFormation.State.collecting,
+                            lambda: self.subbehavior_with_name('pass').state == behavior.Behavior.State.cancelled or \
+                                    self.subbehavior_with_name('pass').state == behavior.Behavior.State.failed,
+                            'Passing: Ball Lost')
+        self.add_transition(
+            AdaptiveFormation.State.shooting,
+            AdaptiveFormation.State.collecting,
+            lambda: self.subbehavior_with_name('kick').is_done_running(),
+            'Shooting: Ball Lost / Shot')
+        self.add_transition(
+            AdaptiveFormation.State.clearing,
+            AdaptiveFormation.State.collecting,
+            lambda: self.subbehavior_with_name('clear').is_done_running(),
+            'Clearing: Ball Lost')
+
+    def should_pass_from_dribble(self):
+
+        # If pass is above cutoff and we dont have a good shot
+        if (self.pass_score > self.dribble_to_pass_cutoff and \
+            self.shot_chance < self.dribble_to_shoot_cutoff):
+            print("Pass : " + str(self.pass_score) + " Shot : " + str(
+                self.shot_chance))
+            return True
+
+        # Force pass if we are near our max dribble dist
+        dribble_dist = (main.ball().pos - self.dribble_start_pt).mag()
+        if (dribble_dist > self.max_dribble_dist):
+            return True
+
+        # Under cutoff
+        return False
+
+    def should_shoot_from_dribble(self):
+
+        # If shot chance is improving significantly, hold off a second
+        if (self.prev_shot_chance + self.IncreasingChancesCutoff <
+                self.shot_chance):
+            return False
+
+        # Not in front of the half
+        if (main.ball().pos.y < constants.Field.Length / 2):
+            return False
+
+        # If shot is above cutoff
+        if (self.shot_chance > self.dribble_to_shoot_cutoff):
+            print("Pass : " + str(self.pass_score) + " Shot : " + str(
+                self.shot_chance))
+            return True
+
+        # Decreasing and under cutoff
+        return False
+
+    def should_clear_from_dribble(self):
+        # If outside clear zone
+        if (self.dribbler.pos.y > self.clear_field_cutoff):
+            return False
+
+        # If pass chances are getting better, hold off
+        if (self.prev_pass_score + IncreasingChancesCutoff < self.pass_score):
+            return False
+
+        # TODO: See if there is space to dribble
+        closest_distance = (evaluation.opponent.get_closest_opponent(
+            main.ball().pos, 1).pos - main.ball().pos).mag()
+        if (closest_distance > self.clear_distance_cutoff):
+            return False
+
+        return True
+
+    def on_enter_collecting(self):
+        # 2 man to man defenders and 1 zone defender
+        defensive_forward = tactics.defensive_forward.DefensiveForward()
```

Comment:
Yeah, that was one thing I was running into. In the best case, I would like to have the robots already moving towards the next pass location or opening up positions on the field by exploiting the defense but I couldn't think of a good score function to use. There are a bunch of possibilities here.

For now, I think I'll just make another play to do something really simple like hold the robots at 50% of the ball Y coordinate or something until they are needed as a pass.